### PR TITLE
Feat/eval suite

### PR DESCRIPTION
**Pull request by tranthanhbinh1**
This pull request introduces significant enhancements to the evaluation and optimization framework, including new evaluation strategies, prompt optimization capabilities, and modular refactoring. Key changes include the addition of evaluation counters, prompt optimization utilities, and a revamped evaluation strategy manager. It also introduces new configuration files and updates existing components for better modularity and functionality.

### Evaluation Enhancements:
* Added `EvaluationCounter` in `run_evaluation.py` to manage persistent evaluation run counts, enabling full or similarity-based evaluations based on thresholds.
* Refactored `Evaluator` in `src/evaluation/eval.py` to support hybrid evaluation strategies and synthetic data generation, replacing legacy evaluation logic.

### Prompt Optimization:
* Introduced `run_prompt_optimization.py` for optimizing prompts using genetic algorithms, configurable strategies, and sentiment-specific evaluators.
* Added `PromptimalEvaluatorAdapter` in `src/evaluation/promptimal_adapter.py` to integrate the `EvaluationStrategyManager` with the `promptimal` library for seamless prompt evaluations.

### Configuration and Strategy Management:
* Added `config/strategy.json` to define evaluation strategies, including similarity, hybrid, and trainer-based approaches, with iteration-specific configurations.
* Updated `src/evaluation/similarity_evaluator.py` (renamed from `prompt_evaluator.py`) to align with the new modular structure and integrate with the `EvaluatorRepository`.

### Miscellaneous Improvements:
* Added `.cursor/rules/pytorch-agent.mdc` with coding guidelines for PyTorch engineers, emphasizing type hints, modularity, and PEP8 compliance.
* Introduced a new workspace configuration in `llm-data-aug.code-workspace` for streamlined project management.
* Added a `SEED` constant in `src/constants.py` for reproducibility.

Disclaimer: This summary is entirely created by.. an LLM!

## Summary by Sourcery

Unify and modularize the evaluation and training framework by extracting common repositories and preprocessing steps, introduce a flexible evaluation strategy manager, and build a prompt optimization pipeline with genetic algorithms and Promptimal integration, complemented by persistent evaluation counters and automated tests.

New Features:
- Add run_prompt_optimization.py script for prompt optimization using genetic algorithms and customizable evaluation strategies
- Introduce EvaluationStrategyManager to orchestrate similarity, trainer, hybrid, and LLM-judge evaluation methods
- Provide PromptimalEvaluatorAdapter for seamless integration with Promptimal optimizer
- Implement persistent EvaluationCounter in run_evaluation.py to switch between full and similarity-based evaluations

Enhancements:
- Refactor LSTM, CNN-BERT hybrid, SVM, and PhoBert trainers to implement a unified TrainerEvaluatorRepository interface
- Centralize text cleaning and tokenization in a new TextPreprocessor class under PreprocessorRepository
- Restructure evaluation code into modular components, replacing legacy prompt evaluator with SimiarityEvaluator and strategy manager
- Clean up utils by removing inline text utilities and adding reusable helper functions
- Add constants (SEED, ORIGINAL_DATASET_PATH) and reorganize import paths for consistency
- Update run_evaluation.py to support dynamic evaluation modes based on global count and threshold configuration

Documentation:
- Add coding guidelines in .cursor/rules/pytorch-agent.mdc
- Include editor workspace settings in llm-data-aug.code-workspace
- Add config/strategy.json to define evaluation switching rules

Tests:
- Add test_promptimal_integration.py to validate Promptimal integration and evaluator creation
- Include basic import, factory, and mock optimization tests for the new evaluation adapters

### CODE REVIEW

File: src/evaluation/strategy_manager.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+        trainer_evaluator: Any,  # TrainerEvaluatorRepository
+        sentiment: str,
+        evaluation_rules: List[EvaluationRule],
+        llm_judge_evaluator: Optional[Callable] = None,
+    ):
+        self.similarity_evaluator = similarity_evaluator
+        self.trainer_evaluator = trainer_evaluator
+        self.sentiment = sentiment
+        self.evaluation_rules = sorted(
+            evaluation_rules, key=lambda r: r.iteration_start
```
[the remark this answers is not in the document]
Comment (tranthanhbinh1) [reply src/evaluation/strategy_manager.py:51]: What do u mean there is no return statement here
